### PR TITLE
[20062] Created TypeObjectRegistryObserver

### PR DESCRIPTION
**Pull request by adriancampo**
## Contributor Checklist
- [x] Commit messages follow the project guidelines. 
- [x] The code follows the style guidelines of this project. 
- N/A Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added; the added tests pass locally 
- N/A] Any new/modified methods have been properly documented using Doxygen. 
- N/A Changes are ABI compatible. 
- N/A Changes are API compatible. 
- N/A New feature has been added to the `versions.md` file (if applicable).
- N/A New feature has been documented/Current behavior is correctly described in the documentation. 
    
- N/A Applicable backports have been included in the description.

## Reviewer Checklist
- [N/A] The PR has a milestone assigned.
- [x] Check contributor checklist is correct.
- [ ] Check CI results: changes do not issue any warning.
- [N/A] Check CI results: failing tests are unrelated with the changes.

### CODE REVIEW

**Review by JLBuenoLopez-eProsima — CHANGES_REQUESTED** [review]

File: test/unittest/rtps/discovery/CMakeLists.txt [hunk 1]
```diff
@@ -80,6 +84,7 @@ target_include_directories(EdpTests PRIVATE
     ${PROJECT_SOURCE_DIR}/test/mock/rtps/TimedEvent
     ${PROJECT_SOURCE_DIR}/test/mock/rtps/ResourceEvent
     ${PROJECT_SOURCE_DIR}/test/mock/rtps/SecurityManager
+    ${PROJECT_SOURCE_DIR}/test/mock/rtps/RTPSDomainImpl
```
[the remark this answers is not in the document]
Comment (adriancampo) [reply test/unittest/rtps/discovery/CMakeLists.txt:87]: Yes, they do

File: test/mock/dds/QosPolicies/fastdds/dds/core/policy/QosPolicies.hpp [hunk 2]
```diff
@@ -1179,6 +1179,9 @@ typedef enum DataRepresentationId : int16_t
     XCDR2_DATA_REPRESENTATION   //!<
 } DataRepresentationId_t;
 
+//! Default @ref DataRepresentationId used in Fast DDS.
+constexpr DataRepresentationId_t DEFAULT_DATA_REPRESENTATION {DataRepresentationId_t::XCDR_DATA_REPRESENTATION};
+
```
[the remark this answers is not in the document]
Comment (adriancampo) [reply test/mock/dds/QosPolicies/fastdds/dds/core/policy/QosPolicies.hpp:1184]: You assume correctly

File: src/cpp/rtps/builtin/discovery/endpoint/EDP.cpp [hunk 3]
```diff
@@ -317,45 +334,63 @@ bool EDP::newLocalWriterProxyData(
...
+                            else if (type_ids.type_identifier1()._d() == fastdds::dds::xtypes::EK_MINIMAL)
+                            {
+                                // TODO (adelcampo) Change to xtypes
+                                //wpd->type_id().m_type_identifier = type_ids.type_identifier1();
+                            }
+                            else
+                            {
+                                // TODO (adelcampo) Change to xtypes
+                                //wpd->type_id().m_type_identifier = type_ids.type_identifier2();
+                            }
```
[the remark this answers is not in the document]
Comment (adriancampo) [reply src/cpp/rtps/builtin/discovery/endpoint/EDP.cpp]: Removed old logic in [1f8992c](https://github.com/eProsima/Fast-DDS/pull/4092/commits/1f8992caa9780a40aeacd9d55cde9fe865f51f07)

File: src/cpp/rtps/builtin/discovery/endpoint/EDP.cpp [hunk 4]
```diff
@@ -182,51 +181,69 @@ bool EDP::newLocalReaderProxyData(
...
+                            else if (type_ids.type_identifier1()._d() == fastdds::dds::xtypes::EK_MINIMAL)
+                            {
+                                // TODO (adelcampo) Change to xtypes
+                                //rpd->type_id().m_type_identifier = type_ids.type_identifier1();
+                            }
+                            else
+                            {
+                                // TODO (adelcampo) Change to xtypes
+                                //rpd->type_id().m_type_identifier = type_ids.type_identifier2();
```
[the remark this answers is not in the document]
Comment (adriancampo) [reply src/cpp/rtps/builtin/discovery/endpoint/EDP.cpp]: Removed old logic in [1f8992c](https://github.com/eProsima/Fast-DDS/pull/4092/commits/1f8992caa9780a40aeacd9d55cde9fe865f51f07)

File: test/mock/rtps/RTPSDomainImpl/rtps/RTPSDomainImpl.hpp [hunk 5]
```diff
@@ -104,6 +105,20 @@ class RTPSDomainImpl
         return RTPSDomain::createParticipant(domain_id, enabled, att, listen);
     }
 
+    static fastdds::dds::xtypes::ITypeObjectRegistry& type_object_registry()
+    {
+        auto instance = get_instance();
+        return instance->type_object_registry_;
```
[the remark this answers is not in the document]
Comment (adriancampo) [reply test/mock/rtps/RTPSDomainImpl/rtps/RTPSDomainImpl.hpp]: [1f8992c](https://github.com/eProsima/Fast-DDS/pull/4092/commits/1f8992caa9780a40aeacd9d55cde9fe865f51f07)

File: src/cpp/fastdds/xtypes/type_representation/TypeObjectUtils.cpp [hunk 6]
```diff
@@ -51,6 +54,14 @@ constexpr const EnumeratedLiteralFlag enum_literal_flag_mask = MemberFlagBits::T
         MemberFlagBits::TRY_CONSTRUCT2 | MemberFlagBits::IS_EXTERNAL | MemberFlagBits::IS_OPTIONAL |
         MemberFlagBits::IS_MUST_UNDERSTAND | MemberFlagBits::IS_KEY;
 
+
+
+
```
[the remark this answers is not in the document]
Comment (adriancampo) [reply src/cpp/fastdds/xtypes/type_representation/TypeObjectUtils.cpp]: Removed in [1f8992c](https://github.com/eProsima/Fast-DDS/pull/4092/commits/1f8992caa9780a40aeacd9d55cde9fe865f51f07)

File: src/cpp/fastdds/xtypes/type_representation/TypeObjectUtils.cpp [hunk 7]
```diff
@@ -21,14 +21,17 @@
...
 #include <fastdds/dds/xtypes/common.hpp>
 #include <fastdds/dds/xtypes/dynamic_types/Types.hpp>
 #include <fastdds/dds/xtypes/exception/Exception.hpp>
 #include <fastdds/dds/xtypes/type_representation/TypeObject.hpp>
 #include <fastdds/dds/log/Log.hpp>
+
 #include <fastrtps/utils/md5.h>
 
+#include <rtps/RTPSDomainImpl.hpp>
+#include <fastdds/xtypes/type_representation/TypeObjectRegistry.hpp>
```
[the remark this answers is not in the document]
Comment (adriancampo) [reply src/cpp/fastdds/xtypes/type_representation/TypeObjectUtils.cpp]: Sorted in [1f8992c](https://github.com/eProsima/Fast-DDS/pull/4092/commits/1f8992caa9780a40aeacd9d55cde9fe865f51f07)

File: src/cpp/fastdds/xtypes/type_representation/TypeObjectUtils.cpp [hunk 8]
```diff
@@ -21,14 +21,17 @@
 
 #include <fastcdr/exceptions/BadParamException.h>
 
-#include <fastdds/dds/domain/DomainParticipantFactory.hpp>
 #include <fastdds/dds/xtypes/common.hpp>
 #include <fastdds/dds/xtypes/dynamic_types/Types.hpp>
 #include <fastdds/dds/xtypes/exception/Exception.hpp>
 #include <fastdds/dds/xtypes/type_representation/TypeObject.hpp>
 #include <fastdds/dds/log/Log.hpp>
+
```
[the remark this answers is not in the document]
Comment (adriancampo) [reply src/cpp/fastdds/xtypes/type_representation/TypeObjectUtils.cpp]: Removed in [1f8992c](https://github.com/eProsima/Fast-DDS/pull/4092/commits/1f8992caa9780a40aeacd9d55cde9fe865f51f07)

File: src/cpp/fastdds/xtypes/type_representation/TypeObjectRegistry.hpp [hunk 9]
```diff
@@ -253,6 +203,76 @@ class TypeObjectRegistry
...
+            uint32_t& type_object_serialized_size);
+
+#ifndef DOXYGEN_SHOULD_SKIP_THIS_PUBLIC
+    // Only DomainParticipantFactory is allowed to instantiate the TypeObjectRegistry class.
+    // It cannot be protected as the standard library needs to access the constructor to allocate the resources.
+    // Rule of zero: resource managing types.
+    TypeObjectRegistry();
+#endif // DOXYGEN_SHOULD_SKIP_THIS_PUBLIC
+
+protected:
```
[the remark this answers is not in the document]
Comment (adriancampo) [reply src/cpp/fastdds/xtypes/type_representation/TypeObjectRegistry.hpp]: Fixed in [1f8992c](https://github.com/eProsima/Fast-DDS/pull/4092/commits/1f8992caa9780a40aeacd9d55cde9fe865f51f07)

File: src/cpp/fastdds/xtypes/type_representation/TypeObjectRegistry.hpp [hunk 10]
```diff
@@ -253,6 +203,76 @@ class TypeObjectRegistry
...
+     *
+     * @param[in] type_object TypeObject which is to be hashed.
+     * @param[out] type_object_serialized_size
+     * @return const TypeIdentifier related with the given TypeObject.
+     */
+    const TypeIdentifier calculate_type_identifier(
+            const TypeObject& type_object,
+            uint32_t& type_object_serialized_size);
+
+#ifndef DOXYGEN_SHOULD_SKIP_THIS_PUBLIC
```
[the remark this answers is not in the document]
Comment (adriancampo) [reply src/cpp/fastdds/xtypes/type_representation/TypeObjectRegistry.hpp]: Fixed in [1f8992c](https://github.com/eProsima/Fast-DDS/pull/4092/commits/1f8992caa9780a40aeacd9d55cde9fe865f51f07)

File: src/cpp/rtps/RTPSDomain.cpp [hunk 11]
```diff
@@ -797,6 +797,18 @@ void RTPSDomainImpl::set_filewatch_thread_config(
     instance->callback_thread_config_ = callback_thread;
 }
 
+fastdds::dds::xtypes::ITypeObjectRegistry& RTPSDomainImpl::type_object_registry()
+{
+    auto instance = get_instance();
+    return instance->type_object_registry_;
```
[the remark this answers is not in the document]
Comment (adriancampo) [reply src/cpp/rtps/RTPSDomain.cpp]: Fixed in [1f8992c](https://github.com/eProsima/Fast-DDS/pull/4092/commits/1f8992caa9780a40aeacd9d55cde9fe865f51f07)

File: src/cpp/fastdds/xtypes/type_representation/TypeObjectUtils.cpp [hunk 12]
```diff
@@ -2015,7 +2007,7 @@ void TypeObjectUtils::struct_member_flag_consistency(
     {
         throw InvalidArgumentError("Keyed members must have their \"must understand\" attribute set to true");
     }
-    if ((MemberFlagBits::IS_DEFAULT & member_flags) != 0)
+    if ((MemberFlagBits::IS_DEFAULT& member_flags) != 0)
```
[the remark this answers is not in the document]
Comment (adriancampo) [reply src/cpp/fastdds/xtypes/type_representation/TypeObjectUtils.cpp]: Undid in [8d756e9](https://github.com/eProsima/Fast-DDS/pull/4092/commits/8d756e9b49ad7bc199c15fd76fbba1f5aaabd660)

File: src/cpp/fastdds/xtypes/type_representation/dds_xtypes_typeobjectCdrAux.ipp
[the remark this answers is not in the document]
Comment (adriancampo) [reply src/cpp/fastdds/xtypes/type_representation/dds_xtypes_typeobjectCdrAux.ipp:1]: Removed in [1f8992c](https://github.com/eProsima/Fast-DDS/pull/4092/commits/1f8992caa9780a40aeacd9d55cde9fe865f51f07)

**Review by JLBuenoLopez-eProsima — APPROVED** [review]
LGTM with green CI (build)

**Pull request merged**